README.md
# ISW / Ice-Sealed Wyvern
<img src="https://github.com/YoyPa/isw/blob/master/image/isw.svg" alt="" width="25%" align="right">

- isw started as an equivalent of "control tools by pherein" but under linux.
- It is meant to alter fan profiles of MSI laptops.
- Profiles for supported laptops are located in <a href="https://github.com/YoyPa/isw/blob/master/etc/isw.conf">/etc/isw.conf</a>.
- You can check <a href="https://github.com/YoyPa/isw/blob/master/etc/isw.conf">/etc/isw.conf</a>. comments for more details.

## Warning
- Use it at your own risk!
- Secure boot can prevent access to the EC.
- isw is made/tested with MSI GS40 6QE under Arch/Manjaro, other laptops depend on user contribution.
- Check that your EC (Embedded Controler) work the same way, you can find documentation on the <a href="https://github.com/YoyPa/isw/wiki/MSI-G-laptop-EC---Rosetta">wiki</a>.

## How to install
### Package or not package ?
- If you are using archlinux or a derivative you can install it from AUR: ```yay -S isw```
- If you are using debian or a derivate:
  - Clone ```git clone https://github.com/YoyPa/isw```
  - Install build tools for packaging ```apt install build-essential devscripts```
  - Build the package ```debuild -us -uc -b```
  - Install it ```apt install ../isw_*.deb```
- If you are on a different distro family:
  - Clone ```git clone https://github.com/YoyPa/isw```
  - Then look at this <a href="https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=isw">PKGBUILD</a> to know where to put files.
  - /!\ Path can change a bit depending on the distro /!\

### Builtin or not builtin ?
- It need ```ec_sys``` module with option ```write_support=1```, there are two scenario to set that:
  - ec_sys is a builtin kernel module:
    - add ```ec_sys.write_support=1``` in ```/etc/default/grub``` (Arch AUR package can't do it for you).
    - then update your grub with ```update-grub``` and reboot.
  - ec_sys is not a builtin kernel module:
    - copy both ```isw-ec_sys.conf``` files provided (/etc/mod[...]) with same path (Arch AUR package will do it for you).
    - then reboot OR ```modprobe ec_sys write_support=1```.

## How to use it ?
### Current --help output
```
usage: isw [-h] [-b B] [-c] [-f FILE] [-p P] [-r [R]] [-s S S] [-t T] [-u USB] [-w W]

optional arguments:
  -h, --help            show this help message and exit
  -b B                  ┬ enable or disable CoolerBoost
                        └ replace B with off OR on

  -c                    ─ show an EC dump

  -f FILE, --file FILE  ┬ show profile in EC update file
                        └ replace FILE with FILE_NAME

  -p P                  ┬ show current profile in EC
                        └ replace P with SECTION_NAME

  -r [R]                ┬ show realtime CPU+GPU temp and fan speed from EC
                        ├ replace [R] with any [NUMBER] to perform a [NUMBER] of time(s)
                        └ Assume [0] if given nothing = infinite loop

  -s S S                ┬ set a single value into EC
                        ├ replace 1st S with ADDRESS in hexadecimal (0x00)
                        └ replace 2nd S with VALUE   in decimal     (00)

  -t T                  ┬ set the battery charging treshold
                        └ replace T with a NUMBER between 20 and 100 (٪)

  -u USB, --usb USB     ┬ set usb backlight level
                        └ replace USB with off, half OR full

  -w W                  ┬ write into EC
                        └ replace W with SECTION_NAME

┌─ TIPS ──────────────────────────────────────────────────────────────────┐
│ Set your config in '/etc/isw.conf'.                                     │
│ Arguments order is relevant, -c and -p can be used twice. Example:      │
│ isw -cw SECTION_NAME -c will show you EC dump before and after change.  │
├─ SUPPORT ───────────────────────────────────────────────────────────────┤
│ Help me support your laptop by providing following command output:      │
│ isw -cp MSI_ADDRESS_DEFAULT                                             │
│ via https://github.com/YoyPa/isw (open an issue).                       │
│ Make sure your dump is made before altering EC with isw, you can reset  │
│ your EC with a reboot or by changing power source.                      │
├─ NAME ──────────────────────────────────────────────────────────────────┤
│ ISW is MSI at 180°                                                      │
│ It means Ice-Sealed Wyvern in opposition to MSI's 'unleash the dragon'  │
└─────────────────────────────────────────────────────────────────────────┘
```
<b>NB: all option exept -h and -f need priviledges.</b>

### An example
<b>SECTION_NAME</b> refer to the motherboard name inside ```isw.conf```, if <b>for example</b> you have a <b>GS40_6QE</b> your <b>SECTION_NAME</b> would be <b>14A1EMS1</b>.

If you want to change temperature treshold and/or fan speed for cpu and/or gpu, you have to edit the corresponding section in ```isw.conf``` to set the wanted values and use ```isw -w 14A1EMS1``` to apply.

If you want to check the current temperature and fan speed you will have to type ```isw -r```.

Don't forget to read the comment at the beginning of ```isw.conf```, it contain some helpfull info.

## Launch at startup/resume
You can launch ```isw -w [SECTION_NAME]``` at startup/resume via systemd with isw@.service (need priviledges):
```
systemctl enable isw@[SECTION_NAME].service
```

## TODO
```
- Daemonisation
	- Launch at startup                            done
	- launch after resume (hibernation/suspend)    done
	- Launch at event(power source change)
```


debian/changelog
isw (1.10-2) unstable; urgency=medium

  * Ship ec_sys as a dkms module

 -- Jose Angel Pastrana Padilla <japp0005@red.ujaen.es>  Sun, 08 Aug 2021 10:45:29 +0200

isw (1.10-1) unstable; urgency=medium

  * Release 1.10

 -- Hayden Hughes <hayden@vulpes.chickenkiller.com>  Wed, 12 Feb 2020 13:10:35 +1100


isw_install.py
#! /usr/bin/python3

from tkinter import *
from tkinter import messagebox
import git
import getpass
import os

username = getpass.getuser()
path_check = "/home/" + username + "/isw"
check = os.path.exists(path_check)
path = "/home/" + username

window = Tk()
window.title('ISW-GUI')

lable1 = Label(window, text = "ISW is opposite of MSI", fg = 'red', font=("Helvetica", 20))
lable1.grid(row = 1, column = 0)
lable2 = Label(window, text = "please enter your laptop model.\nFor model details visit :-> https://github.com/YoyPa/isw", fg = 'black', font=("Helvetica", 16))
lable2.grid(row = 2, column = 0)

laptop_model = Entry(window, font=("Helvetica", 20))
laptop_model.grid(row = 3, column = 0)

def install():
    if check == False:
        git.Git(path).clone("https://github.com/YoyPa/isw")
        messagebox.showinfo('ISW-GUI', 'ISW folder is copied to HOME folder from GIT page!\nSystem will reboot now and then set your laptop model fan speed curve!')
    if check == True:
        messagebox.showinfo('ISW-GUI', 'ISW folder already exist!')
    os.system("gnome-terminal -e 'bash -c \"sudo apt install nautilus-admin && sudo isw -r && install -Dm 644 etc/isw.conf \"${pkgdir}/etc/isw.conf\" && install -Dm 644 etc/modprobe.d/isw-ec_sys.conf \"${pkgdir}/etc/modprobe.d/isw-ec_sys.conf\" && install -Dm 644 etc/modules-load.d/isw-ec_sys.conf \"${pkgdir}/etc/modules-load.d/isw-ec_sys.conf\" && install -Dm 644 usr/lib/systemd/system/isw@.service \"${pkgdir}/usr/lib/systemd/system/isw@.service\" && install -Dm 755 isw \"${pkgdir}/usr/bin/isw\"; exec bash\"'")
    os.system("gnome-terminal -x reboot")
    return

def monitor():
    laptop = laptop_model.get()
    messagebox.showinfo('Message', 'You clicked the monitor button!')
    command = "gnome-terminal -x sudo isw -w " + laptop
    os.system(command)
    os.system("gnome-terminal -x sudo isw -r")
    return

def startup():
    laptop = laptop_model.get()
    command = "gnome-terminal -x systemctl enable isw@" + laptop + ".service"
    messagebox.showinfo('Message', 'You clicked the startup button!')
    os.system(command)
    return

def uninstall():
    messagebox.showinfo('Message', 'You clicked the uninstall ISW button!')
    os.system("gnome-terminal -e 'bash -c \"sudo rm /etc/isw.conf /etc/modprobe.d/isw-ec_sys.conf /etc/modules-load.d/isw-ec_sys.conf /usr/lib/systemd/system/isw@.service /usr/bin/isw; exec bash\"'")
    messagebox.showinfo('ISW-GUI', 'ISW is installed.\nSystem will reboot now and then set your laptop model fan speed curve!')
    os.system("gnome-terminal -x reboot")
    return

button1 = Button(window, text = "Install ISW", width = 20, fg = 'blue', command = install)
button1.grid(row = 4, column = 0)
button3 = Button(window, text = "Monitor Temperature", width = 20, fg = 'green', command = monitor)
button3.grid(row = 5, column = 0)
button4 = Button(window, text = "Run at Startup", width = 20, fg = 'yellow', command = startup)
button4.grid(row = 6, column = 0)
button5 = Button(window, text = "Uninstall ISW", width = 20, fg = 'red', command = uninstall)
button5.grid(row = 7, column = 0)

window.mainloop()


isw.py
#! /usr/bin/python

# IMPORTS
import sys
import os
from optparse import OptionParser

# VARIABLES GLOBALES
EC_IO_FILE="/sys/kernel/debug/ec/ec0/io"
EC_SYS_STARTUP="/etc/modules-load.d/isw-ec_sys.conf"
EC_SYS_STARTUP_OPT="/etc/modprobe.d/isw-ec_sys.conf"
DUMP_POS=""
ADDRESS=0x6a
VALUE=68
X=int(5)

# FONCTIONS
# option -l
def check_ec_sys(option, opt, value, parser):
	if not os.path.exists(EC_IO_FILE):
		os.system("modprobe ec_sys write_support=1")
		print("ec_sys module have been loaded")
	else:
		print("ec_sys module is already loaded")

# option -s
def check_ec_sys_startup(option, opt, value, parser):
	if not os.path.exists(EC_SYS_STARTUP):
		with open(EC_SYS_STARTUP,"xt") as file:
			file.write("ec_sys\n")
		print(EC_SYS_STARTUP+" created")
	else:
		print("nothing to do "+EC_SYS_STARTUP+" already exist")
	if not os.path.exists(EC_SYS_STARTUP_OPT):
		with open(EC_SYS_STARTUP_OPT,"xt") as file:
			file.write("options ec_sys write_support=1\n")
		print(EC_SYS_STARTUP_OPT+" created")
	else:
		print("nothing to do "+EC_SYS_STARTUP_OPT+" already exist")

# option -a
def change_address(option, opt, value, parser):
	global ADDRESS
	ADDRESS=int(value)

# option -v
def change_value(option, opt, value, parser):
	global VALUE
	VALUE=int(value)

# option -x
def change_x(option, opt, value, parser):
	global X
	X=int(value)

# option -w
def ec_write(option, opt, value, parser):
	global ADDRESS
	global VALUE
	global DUMP_POS
	DUMP_POS="after modification"
	print('\r')
	with open(EC_IO_FILE,"wb") as file:
		for i in range(6):
			file.seek(ADDRESS)
			file.write(bytearray((VALUE,)))
			print('Value '+hex(VALUE)+'('+str(VALUE)+'°C)'+' written at '+hex(ADDRESS)+'(byte'+str(ADDRESS)+')')
			ADDRESS=ADDRESS+1
			VALUE=VALUE+X

# option -c
def ec_check(option, opt, value, parser):
	print('\nEC dump '+str(DUMP_POS))
	print('\x1b[0;32;1m' + '       00 01 02 03 04 05 06 07 08 09 0A 0B 0C 0D 0E 0F' + '\x1b[0m')
	os.system("od -A x -t x1z "+str(EC_IO_FILE))

def main():
	# permet l'utilisation d'arguments
	# callback permet l'appel d'une fonction
	parser = OptionParser(epilog=
"""TIPS ------------------------------------------------------------------------
Fan will start spinning at VALUE (°C) and increase speed every XJUMP.		 
EC contain 7 values, 6 of them will be edited, last one is left at 0x64(100).
Please ensure that VALUE+5*XJUMP stay below 0x64(100).						 
For options -a -v -x, decimal and hexadecimal values are allowed.			 
NAME ------------------------------------------------------------------------
ISW is MSI at 180°														 	
ISW mean ice-sealed wyvern in opposition to MSI's "unleash the dragon"		 
-----------------------------------------------------------------------------
""")
	parser.add_option("-l", "--load", action="callback", callback=check_ec_sys,
					  help="look for ec_sys module and load it if needed")
	parser.add_option("-s", "--startup", action="callback", callback=check_ec_sys_startup,
					  help="configure ec_sys module for startup (systemd)")
	parser.add_option("-a", "--address", action="callback", callback=change_address, type="int",
					  help="specify starting address (default: 0x6a)")
	parser.add_option("-v", "--value", action="callback", callback=change_value, type="int",
					  help="specify starting value (°C) (default: 68)")
	parser.add_option("-x", "--xjump", action="callback", callback=change_x, type="int",
					  help="specify by how much value will increment (default: 5)")
	parser.add_option("-w", "--write", action="callback", callback=ec_write,
					  help="write into EC, always keep this option at the end")
	parser.add_option("-c", "--check", action="callback", callback=ec_check,
					  help="show an EC dump, -cw=before -wc=after -cwc=both")
	(options, args) = parser.parse_args()

if __name__ == "__main__":
    main()
